get_http.go

```diff
@@ -293,50 +294,79 @@ func (g *HttpGetter) Get(dst string, u *url.URL) error {
 	// If there is a subdir component, then we download the root separately
 	// into a temporary directory, then copy over the proper subdir.
 	source, subDir := SourceDirSubdir(source)
-	if subDir == "" {
-		var opts []ClientOption
 
-		// Check if the protocol was switched to one which was not configured.
-		//
+	var opts []ClientOption
+
+	// Check if the protocol was switched to one which was not configured.
+	if g.client != nil && g.client.Getters != nil {
+		// We must first use the Detectors provided, because `X-Terraform-Get does
+		// not necessarily return a valid URL. We can replace the source string
+		// here, since the detectors would have been called immediately during the
+		// next Get anyway.
+		source, err = Detect(source, g.client.Pwd, g.client.Detectors)
+		if err != nil {
+			return err
+		}
+
+		protocol := ""
+		// X-Terraform-Get allows paths relative to the previous request too,
+		// which won't have a protocol.
+		if !relativeGet(source) {
+			protocol = strings.Split(source, ":")[0]
+		}
+
 		// Otherwise, all default getters are allowed.
-		if g.client != nil && g.client.Getters != nil {
-			protocol := strings.Split(source, ":")[0]
+		if protocol != "" {
 			_, allowed := g.client.Getters[protocol]
 			if !allowed {
 				return fmt.Errorf("no getter available for X-Terraform-Get source protocol: %q", protocol)
 			}
 		}
+	}
 
-		// Add any getter client options.
-		if g.client != nil {
-			opts = g.client.Options
-		}
+	// Add any getter client options.
+	if g.client != nil {
+		opts = g.client.Options
+	}
 
-		// If the client is nil, we know we're using the HttpGetter directly. In this case,
-		// we don't know exactly which protocols are configued, but we can make a good guess.
-		//
-		// This prevents all default getters from being allowed when only using the
-		// HttpGetter directly. To enable protocol switching, a client "wrapper" must
-		// be used.
-		if g.client == nil {
+	// If the client is nil, we know we're using the HttpGetter directly. In
+	// this case, we don't know exactly which protocols are configured, but we
+	// can make a good guess.
+	//
+	// This prevents all default getters from being allowed when only using the
+	// HttpGetter directly. To enable protocol switching, a client "wrapper" must
+	// be used.
+	if g.client == nil {
+		switch {
+		case subDir != "":
+			// If there's a subdirectory, we will also need a file getter to
+			// unpack it.
+			opts = append(opts, WithGetters(map[string]Getter{
+				"file":  new(FileGetter),
```

Comment:
Thank you for the ping!

The intent for this original check was to totally prevent the `HttpGetter` from allowing cross-protocol switching unless the getter was already supported by a `Client`. I understand the reasoning for the `FileGetter` with a subdirectory component, but I _hate_ this is allowed at all. It has the potential to be dangerous, and generally not desirable from a security standpoint.

Thankfully, I don't think it would cause problems for Nomad, who explicitly disable the `FileGetter` because it's dangerous in their context. They use a `Client` directly. But I could see this causing problems in other contexts. Likely outside of HashiCorp. 

I would much rather this be a new option on the `HttpGetter`, rather than be implicitly allowed by default. Maybe something like:

```go
httpGetter := &HttpGetter{
  AllowFileGetterForSubDirectoryComponent: true,
}
```

What do you think? Would that be an acceptable solution?

Reply:
To be clear, this is more limiting than the current release, since `getSubdir` falls back to all the default getters. We can't really change this in version 1, since it's a breaking change. While `X-Terraform-Get` probably shouldn't have been so liberal in spec, the protocol does define any valid go-getter source string is a valid value, including subdirectories. I think the only valid way to avoid this is to disable `X-Terraform-Get`, which we have an option for already.

Reply:
To be clear `getSubdir` was falling back to all default getters when using the `HttpGetter` directly?

Aside from that clarification: I think I agreed with you, that disabling `X-Terraform-Get` would generally be the "right" solution, as opposed to an additional option.

Reply:
Right now these checks are only run if `subDir == ""`, otherwise we just call [`getSubDir`](https://github.com/hashicorp/go-getter/blob/main/get_http.go#L339), which runs [`Get`](https://github.com/hashicorp/go-getter/blob/c5e9d081082f01577c64ae9af69f06e20a1e3bd0/get_http.go#L495) with the default options, bypassing the forced `http`+`https` getters entirely. This change now at least will force the `http`+`https` getters with subdirectories too.